library/alloc/src/raw_vec.rs

```diff
@@ -386,13 +388,32 @@ impl<T, A: Allocator> RawVec<T, A> {
             return Err(CapacityOverflow.into());
         }
 
+        // Size of allocator's per-allocation overhead we expect
+        // FIXME: maybe two pointers to be on the safe side? It could potentially
+        // be platform-dependent.
+        let alloc_overhead_size = mem::size_of::<usize>();
+
         // Nothing we can really do about these checks, sadly.
         let required_cap = len.checked_add(additional).ok_or(CapacityOverflow)?;
 
-        // This guarantees exponential growth. The doubling cannot overflow
-        // because `cap <= isize::MAX` and the type of `cap` is `usize`.
-        let cap = cmp::max(self.cap * 2, required_cap);
-        let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
+        let alloc_size = required_cap.checked_mul(mem::size_of::<T>()).ok_or(CapacityOverflow)?;
```

Comment:
All these new checked operations are unfortunate, but I'll worry about them only if I get any improvements in the RSS that validates the idea.